### ISSUE TITLE
Upgrading from 6.4.1 to 6.5.X impacts response time and cpu usage

### ISSUE THREAD

**Issue by marceloboeira** [thumbor/thumbor#1102]
We have upgraded our docker image from thumbor `6.4.1` to `6.5.1` yesterday. Both CPU and response time are increasing considerably ever since, that's the only change that happened.

### Expected behaviour
Nothing should dramatically change...

### Actual behaviour
Both CPU usage and response-time went up considerably.

<img width="621" alt="screen shot 2018-07-31 at 11 00 11" src="https://user-images.githubusercontent.com/1898225/43449527-f129c8f6-94b0-11e8-8736-e7fe053190ac.png">

* Orange - time to fetch 
* Blue - p95 of response time
* Yellow dashed - Max response time 
* Red - deployment (first ones to upgrade, second one to downgrade)

Once I've downgraded back to 6.4.1 the response time went down again...

<img width="628" alt="screen shot 2018-07-31 at 10 19 44" src="https://user-images.githubusercontent.com/1898225/43447715-5c625d72-94ac-11e8-9b49-b464fd5ac4e8.png">

* CPU Usage per pod after the deployments... 

The spike matches the deployment time, the CPU usage went down over night because of the less intensive traffic but the response time was still higher (1.5x) the one before.  

### Operating system
Ubuntu 16.04 on K8s

### Your thumbor.conf
```
    import re

    ALLOWED_SOURCES=[
      # https://regexper.com/https://regexper.com/#^https:\/\/*(.+)?hey[.]car\/*(.+)
      re.compile(r'^https://*(.+)?hey[.]car/*(.+)'),
    ]

    MAX_SOURCE_SIZE = 500
    SECURITY_KEY = "REDACTED"
    MAX_WIDTH = 2000
    MAX_HEIGHT = 2000

    # Cache
    MAX_AGE = 365 * 24 * 60 * 60
    MAX_AGE_TEMP_IMAGE = 240

    # Defaults
    QUALITY = 100

    # Engines
    HTTP_LOADER_CURL_ASYNC_HTTP_CLIENT = True

    # Metrics
    METRICS = "thumbor.metrics.statsd_metrics"
    STATSD_PORT = 8125
    STATSD_HOST = "datadog-statsd.kube-system"
    STATSD_PREFIX = "image_proxy_service"

    # GC Options
    GC_INTERVAL = 1800

    # AutoWebp
    AUTO_WEBP = True

    # Sentry
    ERROR_HANDLER_MODULE = "thumbor.error_handlers.sentry"
    SENTRY_DSN = "REDACTED"
```

Any thoughts on what could've caused this? Is there any new config that I need to take into account when upgrading? I've been reading the changelog and haven't found anything clearly related to that.

**Comment by caeugusmao** [comment]
Hi @marceloboeira We are gonna take a look. We have upgraded here, we can check if we had this behavior too. Tks for your report.

**Comment by kkopachev** [comment]
@marceloboeira I noticed you have statsd metrics enabled to datadog. Is there any change in response time for all image formats or there is one in particular?

**Comment by marceloboeira** [comment]
@caeugusmao thanks for looking at it. I have lots of metrics available here if you need.

@kkopachev hmm, we serve mostly jpg so I didn't see any difference for a specific format. Let me show you:

<img width="631" alt="screen shot 2018-08-01 at 09 12 13" src="https://user-images.githubusercontent.com/1898225/43506659-02d88bf4-956b-11e8-80eb-1d687abd9816.png">

The bigger one is jpg, smaller is png.

**Comment by gingerlime** [comment]
@marceloboeira Which specific docker image / library versions are you using? If it's custom made, would you mind sharing your Dockerfile? Could you try with https://github.com/MinimalCompact/thumbor and see if there's a difference? (worth a shot?)

Also, this is [a bit tricky with k8s](https://github.com/MinimalCompact/thumbor/issues/1), but if it fits your use-case, and you are able to use the [thumbor-nginx-proxy](https://hub.docker.com/r/minimalcompact/thumbor-nginx-proxy/) companion, it's a fairly cheap way to cache your results and save on thumbor processing. (of course, this doesn't solve the cpu problem, which is definitely worth investigating, but just worth noting).

**Comment by thor27** [comment]
Hi, 

Unfortunately I was not able to reproduce this issue here. We have made the same kind of upgrade without any impact recently. I've also tried to run some trace but I didn't found anything suspicious.

**Comment by marceloboeira** [comment]
@thor27 thanks for the update. it could be that we use different features of thumbor, I've attached my `thumbor.config` file previously for that reason.  I might try once more the upgrade just to make sure...

**Comment by marceloboeira** [comment]
Once again, I've tried to update Thumbor, to the latest 6.5.2, because I don't want to get locked on updates... but...

<img width="572" alt="screen shot 2018-09-24 at 09 41 23" src="https://user-images.githubusercontent.com/1898225/45941361-1160c680-bfde-11e8-8a60-5568d4e0e336.png">

<img width="579" alt="screen shot 2018-09-24 at 09 41 31" src="https://user-images.githubusercontent.com/1898225/45941360-1160c680-bfde-11e8-9e0d-383e8b51ba7b.png">

**once more both CPU and response time went up ...**

**Comment by kkopachev** [comment]
Since there are multiple releases between https://github.com/thumbor/thumbor/compare/6.4.1...6.5.1 can you try upgrading to 6.4.3?

**Comment by marceloboeira** [comment]
@kkopachev yes, I thought of doing that. I might spare some time this week to figure this, kind of bisect to narrow it down to a specific version.

**Comment by NovaDev94** [comment]
I'm also experiencing the same performance decrease when I upgraded from Thumbor 6.3.1 (with Pillow 3.4.2) => Thumbor 6.5.2 (with Pillow 5.2). The response time doubled after I deployed the newer Thumbor version (it's Oct 19, 3:00 AM on the graph).

![image](https://user-images.githubusercontent.com/6073918/47233514-3ab61c00-d3fd-11e8-94c0-53bdcb6d803a.png)

**Comment by caeugusmao** [comment]
@marceloboeira @NovaDev94,
 
Could you test versions 6.4.2 and 6.4.3 and check if the increase occurred after 6.4.3 version? The Pillow changes are in 6.4.3 version. 

We could not realized this behavior in Globo.com usage.

Tks,

**Comment by marceloboeira** [comment]
@caeugusmao I've prepared the builds on my end, I'll test tomorrow both variants and report back the results. 👍

**Comment by marceloboeira** [comment]
<img width="1120" alt="screenshot 2018-10-26 at 07 06 07" src="https://user-images.githubusercontent.com/1898225/47545647-a44a9480-d8ed-11e8-9114-4db2a32b28c9.png">

Around 6:30 (6.4.1 -> 6.4.2) - Still stable
A bit before 7:00 (6.4.2 -> 6.4.3) - Problematic

**Comment by caeugusmao** [comment]
@marceloboeira 

Thanks for the report!! 

We have made a very simple local benchmark testing using wrk with your conf file and we could not reproduce this behavior. :(

```
((HEAD detached at 6.4.2)) $ wrk -t1 -c1 -d60s "http://localhost:8888/unsafe/1654x/filters:quality(85)/https://hey.car/assets/home/stage.png"
Running 1m test @ http://localhost:8888/unsafe/1654x/filters:quality(85)/https://hey.car/assets/home/stage.png
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   585.07ms   54.59ms 775.76ms   88.24%
    Req/Sec     1.00      0.00     1.00    100.00%
  102 requests in 1.00m, 47.95MB read
Requests/sec:      1.70
Transfer/sec:    817.13KB
```

```
((HEAD detached at 6.4.3)) $ wrk -t1 -c1 -d60s "http://localhost:8888/unsafe/1654x/filters:quality(85)/https://hey.car/assets/home/stage.png"
Running 1m test @ http://localhost:8888/unsafe/1654x/filters:quality(85)/https://hey.car/assets/home/stage.png
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   558.41ms   40.31ms 785.43ms   92.52%
    Req/Sec     1.02      0.14     2.00     98.13%
  107 requests in 1.00m, 50.30MB read
Requests/sec:      1.78
Transfer/sec:    857.19KB
```

Could you test the version 6.4.3 with Pillow 4.3.0 and Pillow 3.x instead of Pillow 5.1.0?  I can't see other reason for response time and cpu usage increase. We can generate two versions for this test, one with Pillow 4.3.0 and other with Pillow 3.x. What do you think about that?

**Comment by gingerlime** [comment]
I also ran some tests using [docker images](https://github.com/MinimalCompact/thumbor) using 6.4.2 (with PIL-3.4.2) and 6.5.2 (with PIL-5.1.0) 

... Sorry, but not sure about PIL/Pillow etc. I extracted the version using `import PIL; PIL.__version__`

Similar results as far as I can tell (?)

```
$ docker run -ti --rm -p 8642:80 minimalcompact/thumbor:6.4.2
---> Starting thumbor solo...

~$ wrk -t1 -c1 -d60s "http://localhost:8642/unsafe/1654x/filters:quality(85)/https://hey.car/assets/home/stage.png"                              
Running 1m test @ http://localhost:8642/unsafe/1654x/filters:quality(85)/https://hey.car/assets/home/stage.png
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   337.23ms   32.57ms 758.61ms   98.88%
    Req/Sec     2.65      0.49     3.00     65.73%
  178 requests in 1.00m, 83.67MB read
Requests/sec:      2.96
Transfer/sec:      1.39MB

$ docker run -ti --rm -p 8652:80 minimalcompact/thumbor:6.5.2
---> Starting thumbor solo...

~$ wrk -t1 -c1 -d60s "http://localhost:8652/unsafe/1654x/filters:quality(85)/https://hey.car/assets/home/stage.png"                              
Running 1m test @ http://localhost:8652/unsafe/1654x/filters:quality(85)/https://hey.car/assets/home/stage.png
  1 threads and 1 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   312.71ms   41.87ms 845.14ms   98.96%
    Req/Sec     2.91      0.31     3.00     91.15%
  192 requests in 1.00m, 90.26MB read
Requests/sec:      3.20
Transfer/sec:      1.50MB
```

**Comment by marceloboeira** [comment]
@caeugusmao @gingerlime thanks for putting the effort!

It could be maybe the way I'm building the Docker images?

https://github.com/hey-car/thumbor/blob/master/Dockerfile


https://github.com/hey-car/thumbor
0.2.1 -> Thumbor 6.4.1
0.2.2 -> Thumbor 6.4.2
0.2.21 -> Thumbor 6.4.3
0.2.3 -> Thumbor 6.5.1
0.2.4 -> Thumbor 6.5.2

The Dockerfile is pretty much the same, just installing thumbor...

**Comment by NovaDev94** [comment]
I run Thumbor directly on Heroku (not under Docker) with these configs. My input and output formats are mostly JPG also (around 95% of the requests).

```
QUALITY = 95
PILLOW_RESAMPLING_FILTER = 'LANCZOS'
PILLOW_PRESERVE_INDEXED_MODE = False
MAX_AGE = 24 * 60 * 60 * 30
ALLOW_ANIMATED_GIFS = False
STORAGE = 'thumbor.storages.no_storage'
USE_CUSTOM_ERROR_HANDLING = True
SENTRY_DSN_URL = '...'
```

Other settings are kept the same. Our team was quite busy lately so we haven't really had the time to do some local benchmarks or version regression yet.

**Comment by gingerlime** [comment]
@marceloboeira If you're already using docker, can you maybe try with `minimalcompact/thumbor:6.5.2` ?

**Comment by gingerlime** [comment]
@marceloboeira @caeugusmao I also tried testing with your docker images (tested tags 0.2.2 and 0.2.21 which you indicated should show a difference, between 6.4.2 and 6.4.3, right?). I couldn't see any difference in performance between those two images.

I checked those two images, and 6.4.2 has PIL 4.3.0, 6.4.3 is using PIL 5.1.0.

Maybe the test case doesn't allow us to reproduce the problem here? @marceloboeira do you have any idea what kind of filters or processing you guys are using typically? I have a feeling we'll need a test case to reproduce the difference in performance somehow... This might also explain why only a few people are experiencing performance issues, but not everyone? (just speculating)

**Comment by gingerlime** [comment]
also @NovaDev94 - if you have any examples of the filters you apply, this can also be useful to try to figure out which type of processing might be slower

**Comment by kkopachev** [comment]
I tried hey.car's images and it seems I can repro the problem. 

6.4.2
```
docker run --rm -it -v thumbor.conf:/config/thumbor.config -p 8642:7000 heycar/thumbor:0.2.2
./wrk -t2 -c5 -d60s "http://localhost:8642/unsafe/480x/filters:quality(90)/https://cdn.hey.car/images/cas/1fc767f58789a7abf28942a0d525df93/original.jpg"
Running 1m test @ http://localhost:8642/unsafe/480x/filters:quality(90)/https://cdn.hey.car/images/cas/1fc767f58789a7abf28942a0d525df93/original.jpg
  2 threads and 5 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   124.12ms    5.32ms 167.69ms   89.86%
    Req/Sec    16.12      4.88    20.00     61.24%
  1933 requests in 1.00m, 73.82MB read
Requests/sec:     32.19
Transfer/sec:      1.23MB
```

6.4.3
```
docker run --rm -it -v thumbor.conf:/config/thumbor.config -p 8642:7000 heycar/thumbor:0.2.21
./wrk -t2 -c5 -d60s "http://localhost:8642/unsafe/480x/filters:quality(90)/https://cdn.hey.car/images/cas/1fc767f58789a7abf28942a0d525df93/original.jpg"
Running 1m test @ http://localhost:8642/unsafe/480x/filters:quality(90)/https://cdn.hey.car/images/cas/1fc767f58789a7abf28942a0d525df93/original.jpg
  2 threads and 5 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   177.46ms   11.94ms 416.57ms   95.27%
    Req/Sec    11.59      3.70    20.00     83.68%
  1352 requests in 1.00m, 51.63MB read
Requests/sec:     22.51
Transfer/sec:      0.86MB
```

Similar results with `-t1 -c1`. 
However png image seems to be doing fine in both versions

**Comment by gingerlime** [comment]
I think I was able to use the config, also with "my" docker image. It's reproducible though with jpeg. I'm getting a difference in performance.

**Comment by kkopachev** [comment]
Yeah, so it seems Pillow 5.0.0 -> 5.1.0.

**Comment by gingerlime** [comment]
I also tested with Pillow-SIMD and I can't see a difference in performance between Pillow-SIMD==4.3.0.post0 and Pillow-SIMD==5.2.0.post0

**Comment by kkopachev** [comment]
I was able to repro with pure Pillow. 
Using this script as a test
```python
from io import BytesIO
from PIL import Image
import time, PIL

def timer(func, label, *args):
    iterations = 5000
    starttime = time.time()
    for x in range(iterations):
        func(*args)
        if time.time()-starttime > 10:
            print("%s: breaking at %s iterations, %.6f per iteration" % (
                label, x+1, (time.time()-starttime)/(x+1.0)))
            break
    if x == iterations-1:
        endtime = time.time()
        print("%s: %.4f s  %.6f per iteration" % (
            label, endtime-starttime, (endtime-starttime)/(x+1.0)))

def run():
    im = Image.open('original.jpg')
    im.load()

timer(run, PIL.__version__)
```
Got these results:
```
5.0.0: breaking at 2038 iterations, 0.004907 per iteration
5.1.0: breaking at 814 iterations, 0.012286 per iteration
```

I'll try to bisect it further.

**Comment by kkopachev** [comment]
Issue affects pillow installed from wheels. Compiled version works fine. 

Also, if compiled against libjpeg-turbo, both versions are able to do ~4150 load operations in 10 seconds. Comparing to ~2000 operations with libjpeg coming from wheels.

**Comment by caeugusmao** [comment]
@kkopachev tks, great analysis!

I installed Pillow from wheels in my local machine which works with osx and even this way I couldn't reproduce. :(

@marceloboeira @NovaDev94 Could you try forcing Pillow 5.0.0 for while? If It works, we can make a downgrade Pillow's version or try to update for newer versions.

**Comment by NovaDev94** [comment]
I tried @kkopachev script on my local machine (Mac OSX) too, Pillow was installed inside a virtualenv, and loading a 3.3MB JPG image from an SSD.

* Installed from wheels
    ```
    4.3.0: breaking at 54 iterations, 0.188555 per iteration
    5.0.0: breaking at 53 iterations, 0.189192 per iteration
    5.1.0: breaking at 54 iterations, 0.186215 per iteration
    ```

* Installed without wheels
    ```
    4.3.0: breaking at 149 iterations, 0.067119 per iteration
    5.0.0: breaking at 148 iterations, 0.067864 per iteration
    5.1.0: breaking at 147 iterations, 0.068367 per iteration
    ```

There are no differences between versions, just the compiled ones are much faster than from-wheels ones. Maybe this is not Pillow' problem but its external dependencies'?

**Comment by kkopachev** [comment]
Since issue is happening on linux machine, it might not be reproducible on osx. 
It is expected, that on same machine there will be performance difference running pillow from wheels and compiled, since wheels contain `libjpeg9` and compiled will be linked against whatever libjpeg you have on your system. On `python:2` docker image libjpeg-turbo is preinstalled, which is by itself faster.

I tested with script above and https://cdn.hey.car/images/cas/1fc767f58789a7abf28942a0d525df93/original.jpg image running inside docker `python:2`

Script mentioned above is in `./t.py`. 

```
wget https://cdn.hey.car/images/cas/1fc767f58789a7abf28942a0d525df93/original.jpg
docker run --rm -it -v $PWD:/usr/src/app/ python:2 bash
root@eed94aeccc6e:/# pip install pillow==5.0.0
Collecting pillow==5.0.0
  Downloading https://files.pythonhosted.org/packages/89/bd/1d9a10f3e8157b7df275740b0782a892a0db387f8286620110c41e5146c7/Pillow-5.0.0-cp27-cp27mu-manylinux1_x86_64.whl (5.8MB)
    100% |████████████████████████████████| 5.9MB 1.4MB/s 
Installing collected packages: pillow
Successfully installed pillow-5.0.0
You are using pip version 18.0, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
root@eed94aeccc6e:/# cd /usr/src/app/
root@eed94aeccc6e:/usr/src/app# python t.py 
5.0.0: breaking at 502 iterations, 0.019947 per iteration
root@eed94aeccc6e:/usr/src/app# python t.py 
5.0.0: breaking at 510 iterations, 0.019649 per iteration
root@eed94aeccc6e:/usr/src/app# 
root@eed94aeccc6e:/usr/src/app# 
root@eed94aeccc6e:/usr/src/app# 
root@eed94aeccc6e:/usr/src/app# pip install pillow==5.1.0
Collecting pillow==5.1.0
  Downloading https://files.pythonhosted.org/packages/00/49/a0483e7308b4b04b5a898789911dbb876d9fea54e7df0453915e47744cfd/Pillow-5.1.0-cp27-cp27mu-manylinux1_x86_64.whl (2.0MB)
    100% |████████████████████████████████| 2.0MB 2.4MB/s 
Installing collected packages: pillow
  Found existing installation: Pillow 5.0.0
    Uninstalling Pillow-5.0.0:
      Successfully uninstalled Pillow-5.0.0
Successfully installed pillow-5.1.0
You are using pip version 18.0, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
root@eed94aeccc6e:/usr/src/app# python t.py 
5.1.0: breaking at 335 iterations, 0.029945 per iteration

**Comment by kkopachev** [comment]
I opened [an issue](https://github.com/python-pillow/Pillow/issues/3441) on Pillow side.
You guys should try building pillow with libjpeg-turbo. That should have visible win.

**Comment by heynemann** [comment]
As usual incredible work by Thumbor's community! Really proud to be a part of a group of so many talented and smart individuals. Thanks for the help everyone!

Do you think we can close this issue, then? We can add to the release notes in the next release a note that people should be building pillow with libjpeg-turbo. What do you think?

**Comment by marceloboeira** [comment]
Thanks a lot for the warm support guys!

**Comment by gingerlime** [comment]
@kkopachev deserves a medal :medal_sports: 

@kkopachev - What's the best workaround for now for docker-based images in your opinion? pinning Pillow to 5.0.0? Compiling it into the docker image? (if so, would appreciate any tips...)

**Comment by kkopachev** [comment]
I'd say compile on image build time with libjpeg-turbo. Library and headers are probably already preinstalled on debian.
We use mozjpeg in our project, though, but that might create different images.

```Dockerfile
RUN PILLOW_VERSION=$(python -c 'import PIL; print(PIL.__version__)') && \
    pip uninstall -y pillow || true && \
# https://github.com/python-pillow/Pillow/pull/3241
    LIB=/usr/lib/x86_64-linux-gnu/ \
# https://github.com/python-pillow/Pillow/pull/3237 or https://github.com/python-pillow/Pillow/pull/3245
    INCLUDE=/usr/include/x86_64-linux-gnu/ \
    pip install --no-cache-dir -U --force-reinstall --no-binary=:all: "pillow<=$PILLOW_VERSION" \
#    --global-option="build_ext" --global-option="--debug" \
    --global-option="build_ext" --global-option="--enable-imagequant" \
    --global-option="build_ext" --global-option="--enable-lcms" \
    --global-option="build_ext" --global-option="--enable-zlib" \
    --global-option="build_ext" --global-option="--enable-jpeg" \
    --global-option="build_ext" --global-option="--enable-tiff"
```

Above should work on base `python:2` image without installing any additional libraries.

For mozjpeg, install it first, then pass additional `PKG_CONFIG_PATH=/opt/mozjpeg/lib64/pkgconfig/` to install line.

**Comment by gingerlime** [comment]
Thanks again, @kkopachev. I'm getting an error related to `imagequant` however, and not sure which dependency is missing

```
    The headers or library files could not be found for imagequant,
    which was requested by the option flag --enable-imagequant
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-TOkji8/pillow/setup.py", line 815, in <module>
        raise DependencyException(msg)
    __main__.DependencyException:
    
    The headers or library files could not be found for imagequant,
    which was requested by the option flag --enable-imagequant   
```

Also, curious about the trade-offs regarding `mozjpeg` that you mentioned. Is it worth using by-default? or which cases are not suitable for it? I think I played around with it using the jpegtran optimizer, but didn't see a big difference, but it wasn't a comprehensive test (maybe even I didn't know how to set it up properly).

**Comment by kkopachev** [comment]
Oh, I build libimagequant from source
```Dockerfile
# Build pngquant
RUN cd ~ && \
	git clone --branch 2.12.1 --recursive https://github.com/kornelski/pngquant.git && \
	cd pngquant/ && \
    ./configure --with-openmp && make && make install && \
    cd lib && make install && \
    # clean up
    cd ~ && rm -rf pngquant/ && \
    ldconfig
```

Or you might want to skip it. It is not distributed with Pillow wheels anyway due to GPLv3 license. It's only useful (IIRC) for paletted pngs when you want to preserve them paletted.

**Comment by kkopachev** [comment]
As for mozjpeg. At first I was thinking that using jpegtran from mozjpeg is enough and it's all it can offer. However, jpegtran is lossless tool and from my understanding it could only optimize Huffman trees a bit and make file progressive with interesting 10 scans ([green martians](https://cloudinary.com/blog/progressive_jpegs_and_green_martians)). 

However, mozjpeg [has more](https://github.com/mozilla/mozjpeg/blob/master/README-mozilla.txt#L4-L9) to [offer](https://calendar.perfplanet.com/2014/mozjpeg-3-0/). 
From what I understand, output image may look different from what would libjpeg(-turbo) produce, but we didn't notice any drawbacks in web use. 
It usually outputs 10 scans with Luma-only scan first, and then separate chroma scans, so image looks weird as it loads. Pillow does not provide a way to customize that behavior (although mozjpeg allows), so I ended up using jpegtran [with scans option](https://github.com/thumbor/thumbor/pull/1094) to re-save with ["steep progressive" ](https://cloudinary.com/blog/progressive_jpegs_and_green_martians#another_scan_script). There is a rumor that progressive jpeg takes more time to decode (times number of scans) and it's disabled on mobile devices, but I am not sure about that.

**Comment by gingerlime** [comment]
Thanks a lot, @kkopachev!

The complexity of image processing never stops to amaze me :) so many variations and options it's incredible. Making informed decisions about those can be tricky...

I think I'll stick to more basic options for the base docker image, but of course if someone needs the ability to tweak things further, they can.

I really appreciate the time and energy you put into explaining things, linking to resources, providing snippets. It's amazing. I am truly grateful. 

p.s. I'd love to add you as contributor to https://github.com/MinimalCompact/thumbor (I think I invited you, but not sure if you received it). I understand if you already have too much on your plate obviously.

**Comment by kkopachev** [comment]
@marceloboeira did you try libjpeg-turbo yet?

**Comment by marceloboeira** [comment]
@kkopachev nope, I might try until the end of the week.

**Comment by marceloboeira** [comment]
@kkopachev sorry for taking so long,... I've tried installing libjpeg-turbo and Pillow before installing thumbor, but I couldn't see much difference, checkout the docker-image:

```Docker
FROM ubuntu:16.04

MAINTAINER heycar Engineering <team-engineering@hey.car>

# Install dependencies
RUN apt-get update && apt-get install -y --no-install-recommends \
    build-essential \
    libcurl4-openssl-dev libssl-dev curl \
    python-dev python-pycurl python-pip python-numpy python-opencv \
    webp libpng-dev libtiff-dev libjasper-dev libjpeg-turbo-progs \
    libdc1394-22-dev libdc1394-22 libdc1394-utils \
    gifsicle libgif-dev && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

# Install thumbor
RUN pip install --upgrade pip
RUN pip install setuptools
RUN pip install Pillow==5.4.0
RUN pip install thumbor==6.6.0

# Mount the config folder
VOLUME /config/

EXPOSE 7000

CMD /usr/local/bin/thumbor --port=7000 -c /config/thumbor.config
```

During the installation of thumbor, I've notice that it uninstalls and installs again Pillow... not sure why tho.

**Comment by marceloboeira** [comment]
@kkopachev I've tried this way:

```
FROM ubuntu:16.04

MAINTAINER heycar Engineering <team-engineering@hey.car>

# Install dependencies
RUN apt-get update && apt-get install -y --no-install-recommends \
    build-essential \
    libcurl4-openssl-dev libssl-dev curl \
    python-dev python-pycurl python-pip python-numpy python-opencv \
    webp libpng-dev libtiff-dev libjasper-dev libjpeg-dev \
    libdc1394-22-dev libdc1394-22 libdc1394-utils \
    gifsicle libgif-dev && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

# Install thumbor
RUN pip install --upgrade pip
RUN pip install setuptools
RUN pip install thumbor==6.6.0
RUN pip install Pillow==5.0.0

# Mount the config folder
VOLUME /config/

EXPOSE 7000

CMD /usr/local/bin/thumbor --port=7000 -c /config/thumbor.config
```

My results were fine in regards to response time (I couldn't notice any reasonable changes), **but** CPU went slightly up, 30/40%... (same amount of machines).

**Comment by kkopachev** [comment]
@marceloboeira When installing Pillow, you should force pip to build binaries. I suspect it still uses supplied binaries. Try installing thumbor first and then use snippet from https://github.com/thumbor/thumbor/issues/1102#issuecomment-433796334 to reinstall pillow building all the libs.

**Comment by marceloboeira** [comment]
@kkopachev I've tried, and I run into the 

```
The headers or library files could not be found for lcms,
    which was requested by the option flag --enable-lcms
```

--------

@gingerlime I see your point, and I believe it makes sense, but at this point I wanted to just be able to update thumbor to the latest so I could buy some time to explore other options. Ideally, we could have an official supported docker image for thumbor on this repo, but I guess it's somewhere complex given the different use cases and dependencies. 

I'll give it a try to the minimal compact image, the last time I've checked there were some differences on the volumes code from the way we've been building the config file and backing into the image, but we've changed that, now we're mounting the volume on runtime with  Kubernetes, so it might be easier to do that now.
